### PR TITLE
Replace HIP_LOCK_ARRAYS macros by functions

### PR DESCRIPTION
**Pull request by masterleinad**
Corresponding to https://github.com/kokkos/kokkos/pull/5770. Analogous to https://github.com/desul/desul/pull/64.

### CODE REVIEW

**Comment by trws** [comment]
Looks good to me pending checks that the RDC/non-RDC issue doesn’t come back.

**Comment by masterleinad** [comment]
> Looks good to me pending checks that the RDC/non-RDC issue doesn’t come back.

The LAMMPS test that failed previously with CUDA appears to run fine with this change.

**Review by dalg24 — DISMISSED** [review]
LGTM but I would like a review from @Rombur

**Comment by masterleinad** [comment]
> The tests are not passing

Yeah, the update in the `fedora` CI containers broke it. Since the fix is only in `develop`, I decided to use that for testing now. Instead, we could also continue to test with `Kokkos` 3.4.01` and disable the `fedora:rawhide` tests.

**Review by Rombur — DISMISSED** [review]

File: .github/workflows/continuous-integration-workflow.yml [hunk 1]
```diff
@@ -35,7 +35,7 @@ jobs:
           git submodule update
       - name: Install Kokkos
         run: |
-          git clone --branch 3.4.01 --single-branch --depth 1 https://github.com/kokkos/kokkos.git kokkos_src
+          git clone --branch develop --single-branch --depth 1 https://github.com/kokkos/kokkos.git kokkos_src
```
Comment (dalg24) [inline .github/workflows/continuous-integration-workflow.yml:38]: Pin the version.

**Review by dalg24 — CHANGES_REQUESTED** [review]

**Review by dalg24 — CHANGES_REQUESTED** [review]
Open another pull request where you bump the Kokkos requirements and update the CI.
Make sure you also update https://github.com/desul/desul/blob/f63c5ee63a2f3aababaad995561d15b2b4951611/atomics/performance_tests/kokkos_based/CMakeLists.txt#L2

**Comment by masterleinad** [comment]
> Open another pull request where you bump the Kokkos requirements and update the CI.

See #90.

**Comment by dalg24** [comment]
Will need to re-run tests after #92 is merged

**Review by dalg24 — DISMISSED** [review]

**Review by dalg24 — APPROVED** [review]

**Review by Rombur — APPROVED** [review]

**Pull request merged**